tests/test_optics.py

```diff
@@ -0,0 +1,181 @@
+
+    # gaussian and shear
+    model = piff.Optical(rzero=0, sigma=sigma, g1=g1, g2=g2, optical_template='des')
+    star = model.draw(star)
+    params = gaussian.fit(star).fit.params
+    np.testing.assert_almost_equal(params[0], sigma, 5)
+    np.testing.assert_almost_equal(params[1], g1, 5)
+    np.testing.assert_almost_equal(params[2], g2, 5)
+
+    # now gaussian, shear, aberration, rzero
+    star = make_empty_star(params=[0.5, 0.8, -0.7, 0.5, -0.2, 0.9, -1, 2.0])
+    model = piff.Optical(rzero=0.1, sigma=sigma, g1=g1, g2=g2, optical_template='des')
+    star = model.draw(star)
+
+def test_disk():
+    print('test read/write')
+    # save and load
+    rzero = 0.1
+    sigma = 1.2
+    g1 = -0.1
+    g2 = 0.05
+    model = piff.Optical(rzero=rzero, sigma=sigma, g1=g1, g2=g2, lam=700.0, optical_template='des')
+    model_file = os.path.join('output','optics.fits')
+    with fitsio.FITS(model_file, 'rw', clobber=True) as f:
+        model.write(f, 'optics')
+        model2 = piff.Optical.read(f, 'optics')
+
+    assert model.kwargs['lam'] == model2.kwargs['lam'],'lam mismatch'
+    assert model.optical_psf_kwargs['lam'] == model2.optical_psf_kwargs['lam'],'optical_psf_kwargs lam mismatch'
+    assert model.optical_psf_kwargs['diam'] == model2.optical_psf_kwargs['diam'],'diam mismatch'
+    assert model.lam_over_r0 == model2.lam_over_r0,'lam_over_r0 mismatch'
+    assert model.kwargs['rzero'] == model2.kwargs['rzero'],'rzero mismatch'
+    assert model.kwargs['sigma'] == model2.kwargs['sigma'],'sigma mismatch'
+    assert model.kwargs['g1'] == model2.kwargs['g1'],'g1 mismatch'
+    assert model.kwargs['g2'] == model2.kwargs['g2'],'g2 mismatch'
+
+#####
+# convenience functions
+#####
+
+def plot_star(star):
+    # convenience function
+    import matplotlib.pyplot as plt
+    plt.figure()
+    plt.imshow(star.data.data)
+    plt.colorbar()
+    plt.show()
+
+def plot_param(rzero=0.1, params=[], g1=0, g2=0, sigma=0):
+    # convenience function
+    model = piff.Optical(rzero=rzero, g1=g1, g2=g2, sigma=sigma, optical_template='des')
+    star = make_empty_star(params=params)
+    star = model.draw(star)
+    plot_star(star)
+    return star.data.data
+
+def make_empty_star(icen=500, jcen=700, ccdnum=28, params=None,
+                    properties={},
+                    fit_kwargs={}):
+
+    properties['ccdnum'] = ccdnum
+    # setting scale is crucial
+    star = piff.Star.makeTarget(x=icen, y=jcen, properties=properties,
+                                scale=0.263)
+
+    if np.shape(params) == ():
+        starfit = None
+    else:
+        starfit = piff.StarFit(params, **fit_kwargs)
+
+    star = piff.Star(star.data, starfit)
+
+    return star
+
+if __name__ == '__main__':
+    test_init()
+    test_optical()
+    test_pupil_im(pupil_path='optics_test/DECam_pupil_128.fits')
+    test_pupil_im(pupil_path='optics_test/DECam_pupil_512.fits')
+    test_kolmogorov()
+    test_shearing()
+    test_gaussian()
+    test_disk()
```

Comment:
I think I will wait for the knn to be merged in, because in going about this I discovered it currently only makes sense in the context of the decam wavefront...

maybe it's a sign that I should 'upgrade' the other components (gaussian, atmosphere, shearing) to parameter (instead of kwarg) status